ktor-client/ktor-client-java/jvm/src/io/ktor/client/engine/java/JavaHttpRequestBodyPublisher.kt

```diff
@@ -0,0 +1,175 @@
+        private var writeInProgress = false
+
+        @Volatile
+        private var done = false
+
+        override fun request(n: Long) {
+            if (done) {
+                return
+            }
+            if (n < 1) {
+                val ex = IllegalArgumentException(
+                    "$subscriber violated the Reactive Streams rule 3.9 by requesting "
+                        + "a non-positive number of elements."
+                )
+                signalOnError(ex)
+            } else {
+                try {
+                    // As governed by rule 3.17, when demand overflows `Long.MAX_VALUE` we treat the signalled demand as
+                    // "effectively unbounded"
+                    outstandingDemand.getAndUpdate { initialDemand: Long ->
+                        if (Long.MAX_VALUE - initialDemand < n) {
+                            Long.MAX_VALUE
+                        } else {
+                            initialDemand + n
+                        }
+                    }
+                    synchronized(this) {
+                        if (!writeInProgress) {
+                            writeInProgress = true
+                            readData()
+                        }
+                    }
+                } catch (e: Exception) {
+                    signalOnError(e)
+                }
+            }
+        }
+
+        override fun cancel() {
+            synchronized(this) {
+                if (!done) {
+                    done = true
+                    closeChannel()
+                }
+            }
+        }
+
+        private fun readData() {
+            // It's possible to have another request for data come in after we've closed the file.
+            if (inputChannel.isClosedForRead) {
+                return
+            }
+
+            launch {
+                do {
+                    val buffer = ByteBuffer.allocate(DEFAULT_BUFFER_SIZE)
+                    val result = try {
+                        inputChannel.readAvailable(buffer)
+                    } catch (e: Throwable) {
+                        signalOnError(e)
+                        closeChannel()
+                        return@launch
+                    }
+
+                    if (result > 0) {
+                        buffer.flip()
+                        signalOnNext(buffer)
+                    }
+                    // If we have more permits, queue up another read.
+                } while (outstandingDemand.decrementAndGet() > 0)
+
+                if (inputChannel.isClosedForRead) {
+                    // Reached the end of the channel, notify the subscriber and cleanup
+                    signalOnComplete()
+                    closeChannel()
+                }
+
+                synchronized(this) {
+                    writeInProgress = false
+                }
+            }
+        }
+
+        private fun closeChannel() {
+            try {
+                inputChannel.cancel()
```

Comment:
Why do we `cancel` channel instead of closing? It can lead to losing some bytes

Reply:
Since it's `ByteReadChannel` it has only `#cancel` method. Should we `#discard` remaining data?